test/infrastructure/vcsim/controllers/envsubst_controller.go

```diff
@@ -0,0 +1,252 @@
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the EnvSubst instance
+	envSubst := &vcsimv1.EnvSubst{}
+	if err := r.Client.Get(ctx, req.NamespacedName, envSubst); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the VCenter instance
+	if envSubst.Spec.VCenter == "" {
+		return ctrl.Result{}, errors.New("Spec.VCenter cannot be empty")
+	}
+
+	vCenter := &vcsimv1.VCenter{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: envSubst.Namespace,
+		Name:      envSubst.Spec.VCenter,
+	}, vCenter); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get VCenter")
+	}
+	log = log.WithValues("VCenter", klog.KObj(vCenter))
+
+	// Fetch the ControlPlaneEndpoint instance
+	if envSubst.Spec.Cluster.Name == "" {
+		return ctrl.Result{}, errors.New("Spec.Cluster.Name cannot be empty")
+	}
+
+	controlPlaneEndpoint := &vcsimv1.ControlPlaneEndpoint{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: envSubst.Namespace,
+		Name:      envSubst.Spec.Cluster.Name,
+	}, controlPlaneEndpoint); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get ControlPlaneEndpoint")
+	}
+	log = log.WithValues("ControlPlaneEndpoint", klog.KObj(controlPlaneEndpoint))
+
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(envSubst, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always attempt to Patch the EnvSubst object and status after each reconciliation.
+	defer func() {
+		if err := patchHelper.Patch(ctx, envSubst); err != nil {
+			log.Error(err, "failed to patch EnvSubst")
+			if reterr == nil {
+				reterr = err
+			}
+		}
+	}()
+
+	// Handle deleted EnvSubst
+	if !controlPlaneEndpoint.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, envSubst, vCenter, controlPlaneEndpoint)
+	}
+
+	// Handle non-deleted EnvSubst
+	return ctrl.Result{}, r.reconcileNormal(ctx, envSubst, vCenter, controlPlaneEndpoint)
+}
+
+func (r *EnvSubstReconciler) reconcileNormal(ctx context.Context, envSubst *vcsimv1.EnvSubst, vCenter *vcsimv1.VCenter, controlPlaneEndpoint *vcsimv1.ControlPlaneEndpoint) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling VCSim ControlPlaneEndpoint")
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if r.sshKeys == nil {
+		r.sshKeys = map[string]string{}
+	}
+
+	key := klog.KObj(vCenter).String()
+	sshKey, ok := r.sshKeys[key]
+	if !ok {
+		bitSize := 4096
+
+		privateKey, err := generatePrivateKey(bitSize)
+		if err != nil {
+			return errors.Wrapf(err, "failed to generate private key")
+		}
+
+		publicKeyBytes, err := generatePublicKey(&privateKey.PublicKey)
+		if err != nil {
+			return errors.Wrapf(err, "failed to generate public key")
+		}
+
+		sshKey = string(publicKeyBytes)
+		r.sshKeys[key] = sshKey
+		log.Info("Created ssh authorized key")
+	}
+
+	// Common variables (used both in supervisor and legacy mode)
```

Comment:
```suggestion
	// Common variables (used both in supervisor and govmomi mode)
```
(similar in other cases)

Reply:
Fixed everywhere